titan-compiler/checker.lua

```diff
@@ -0,0 +1,55 @@
+local checker = {}
+
```

Comment:
I like how short this file got and how we avoided most of the boilerplate. But I suspect that we can improve the traverse function:

* Traverse should be a standalone function instead of a method in the ast nodes. The method on the nodes should be as simple/generic as possible. (When I am doing things like this in Ocaml the nodes have an "fmap" method and then I have separate functions for iteration, folding, etc).

* The traverse function should receive just the node as its parameters. Additional parameters such as `st` should be passed as upvalues.

* The default case of "if none of the above continue iterating" should be built-in to the traverse function.

* The constant "if not ok then return false, err" blocks add a lot of clutter. Traverse should be able to handle this somehow.

Tomorrow I can give a more concrete suggestion of how this might look like.

Reply:
We can have a clear API with we just implement the __pairs metamethod.